### ISSUE TITLE
RAM increases drastically when updating BindableCollection.

### ISSUE THREAD

**Issue by NikasZalias** [Caliburn-Micro/Caliburn.Micro#628]
I have a list of 100 items. I have to refresh my list every 1min. Every time I update the list - RAM increases.
`
private BindableCollection<MyObject> _list = new BindableCollection<MyObject>()`

As I said every 1min I have to update my list, so I have an event that fires every 1 minute.
```

private void OnListChanged(List<MyObject> list) {
    _list.Clear();
    _list.Addrange(list);
}
```
After an event occurs, RAM increases about 10-20mb. The problem is that RAM is never decreased... Looks like GC does not collect the garbage.

I tried to update show only the first 10 items and update only first 10 items, RAM was still increasing, but slower.

For view, I'm using ListView:

```
<ListView
    BorderThickness="0"
    ItemTemplate="{StaticResource ItemTemplate}"
    ItemsSource="{Binding ItemView}"
    Focusable="False" />
```
I tried like that:

```
<ListView
    ScrollViewer.CanContentScroll="True"
    VirtualizingPanel.CacheLength="15"
    VirtualizingPanel.IsVirtualizing="True"
    VirtualizingPanel.CacheLengthUnit="Item"
    VirtualizingPanel.VirtualizationMode="Recycling"
    BorderThickness="0"
    ItemTemplate="{StaticResource ItemTemplate}"
    ItemsSource="{Binding ItemView}"
    Focusable="False" />
```
But Virtualization didn't help either...

What I saw that if I comment listview before running code. RAM doesn't increase anymore, so It's definitely something bad with Bindings...

Could someone please help me, explain to me where is the problem?

P.S. If it's relevant I'm using Caliburn.Micro.

EDIT

P.P.S. I've tried to run GC manually, but nothing happened.

P.P.P.S. I've tried to monitor with memory analyzers and everyone shows me that it's unmanaged memory, and I don't understand what to do next...

EDIT

What I found:

Yes, using List.Clear() clears the references to all objects in the list and sets the Count property to 0. It does not release the reference to the underlying array that stores the object references. You can set the Capacity property to 0 after clearing it to release the reference to that array.

If you actually have OOM problems then doing this is not exactly a cure, it can create more address space fragmentation trouble. Check this answer for hints on how to avoid creating too much garbage in the Large Object Heap.

But BindableCollection doesn't have the capacity or possibility to clear the capacity.

**Comment by nigel-sampson** [comment]
Just confirming that replacing with `BindableCollection<T>` with the out of the box `ObservableCollection<T>` from the framework you're seeing the same problems?

**Comment by NikasZalias** [comment]
@nigel-sampson Yes, replacing to `ObservableCollection<T>` makes the same problem.

**Comment by NikasZalias** [comment]
I noticed, that using `ListBox` instead of `ListView` RAM increases slower...

**Comment by NikasZalias** [comment]
@nigel-sampson I actually managed to fix my problem... Well... Sort of fix it... Not very proud of it, but it works now...
So what I had to do is to change my BindableCollection to List.

**PREVIOUS**
```
 private readonly CollectionViewSource _listViewSource = new CollectionViewSource();
 private readonly BindableCollection <MyObject> _list = new BindableCollection <MyObject>();

 public MyViewModel() 
 {
     _listViewSource.Source = _list;
 }

 public ICollectionView ListView => _listViewSource.View

 private void OnListChanged(List<MyObject> list) {
    _list.Clear();
    _list.Addrange(list);
 }
```

**AFTER**
```
 private CollectionViewSource _listViewSource = new CollectionViewSource();
 private List<MyObject> _list = new List<MyObject>();
 private ICollectionView _listView;

 public MyViewModel()  { }

 public ICollectionView ListView
 {
     get => _listView;
     set => Set(ref _listView, value);
 }

 private void OnListChanged(List<MyObject> list) {
    _list.Clear();
    _list.TrimExcess();
    _list.Addrange(list);
    _listViewSource.Source = _list;
    ListView = _listViewSource.View;
 }
```

I know it's not the most elegant solution, but I'm open to hearing other proposals.

**Comment by NikasZalias** [comment]
Damn... I forgot about UI update... with "this fix" UI is not updated... That's why RAM does not increase anymore.

**Comment by NikasZalias** [comment]
**UPDATE**

I unsubscribed from all the events I was subscribed to. The issue still remains the same.

**Comment by nigel-sampson** [comment]
Sorry to hear that, I don't have any quick fixes.

What about replacing the list entirely freeing up the old list to be garbage collected.

**Comment by NikasZalias** [comment]
Could you collaborate more on that? Just make mutable and then instead of `Clear` and `AddRange` make `new BindableCollection<T>(newList)`? If yes, then how should I notify `public ICollectionView ListView`?
